### ISSUE TITLE
faas-cli build forcibly converts line endings to CRLF 

### ISSUE THREAD

**Issue by ImVexed** [openfaas/faas#653]
On Windows using faas-cli new to create the default Go template and then subsequently building it with faas-cli build without modifying anything fails with:

>  Run "gofmt -s -w" on your Golang code
> The command '/bin/sh -c test -z "$(gofmt -l $(find . -type f -name '*.go' -not -path "./vendor/*" -not -path "./function/vendor/*"))" || { echo "Run \"gofmt -s -w\" on your Golang code"; exit 1; }' returned a non-zero code: 1
> 2018/04/16 13:59:05 ERROR - Could not execute command: [docker build -t go-hi .

This issue was referenced here: https://github.com/openfaas/faas/issues/607 , however, this issue is experienced even without Git.  

## Expected Behaviour
faas-cli build to either function with CRLF or not forcibly convert to CRLF.

## Current Behaviour
faas-cli build forcibly converts the Go files to CRLF causing the gofmt step to fail.

## Possible Solution
Use a text editor like Notepad++ to convert all Go files in the generated build folder to LF, then manually build using Docker instead of faas-cli. 

## Steps to Reproduce (for bugs)
1. faas-cli new go-test --lang go
2. faas-cli build -f ./go-test.yml

## Your Environment
* Docker version `docker version` (e.g. Docker 17.0.05 ):
Docker 18.03.0-ce-win59
* Are you using Docker Swarm or Kubernetes (FaaS-netes)?
Docker Swarm
* Operating System and version (e.g. Linux, Windows, MacOS):
Windows 10
* Link to your project or a code example to reproduce issue:
Default Go template

**Comment by alexellis** [comment]
Hi @ImVexed, thanks for reporting an issue. There are several people using Windows in the community who may be able to advise you. 

Can you elaborate on the point where you say the CLI "forcibly converts" your line-feeds? I'm not aware of code that would do this in the code-base.

@johnmccabe @JockDaRock could either of you take a look?

**Comment by johnmccabe** [comment]
Yeah the CLI shouldn’t alter any of your line endings, usually it’s either the editor being used or git autocrlf config.

Have you taken a look at #607 (an issue search should have returned that issue for you).

**Comment by ImVexed** [comment]
Interesting, it's still failing to build however I'm getting varied results from machine to machine, as far as handler.go being forcibly converted to CRLF goes.

Step 1: faas-cli new go-hi --lang go
[faas-cli new.zip](https://github.com/openfaas/faas/files/1917876/faas-cli.new.zip)
Step 2: faas-cli build -f ./go-hi.yml
[faas-cli build.zip](https://github.com/openfaas/faas/files/1917875/faas-cli.build.zip)
Step 3: faas-cli build -f ./go-hi.yml (However handler.go is converted to only LF pre-build)
[faas-cli build w. LF.zip](https://github.com/openfaas/faas/files/1917877/faas-cli.build.w.LF.zip)

It seems handler.go is having LF preserved, however, main.go is being overwritten each time with CRLF which is causing the build to fail.

**Comment by alexellis** [comment]
Hi @iyovcheva - can you update with your instructions for reproducing the issue on Windows?

**Comment by iyovcheva** [comment]
@ImVexed @alexellis I reproduced the issue with the following:
1. Deployed OpenFaaS on Windows 10
2. `faas-cli template pull`
3. `faas-cli new go-test --lang go`
4. `faas-cli build -f ./go-test.yml`

After it failed with the same error from the issue description, I was testing with `docker build -t go-test -f Dockerfile .`  trying some workarounds in the `Dockerfile`, but don't have a solution to share yet.

**Comment by alexellis** [comment]
Pinging @johnmccabe for input

**Comment by johnmccabe** [comment]
@iyovcheva what is your gitconfig, from the same console as you are running `faas-cli` above.
```
git config --get-all core.autocrlf
```

@alexellis I hadn't realised the dependency on a git binary being on the path had been introduced, I'm assuming its picking up the system/global config and converting on clone (urgh). If that **is** the case you could set the local `core.autocrlf` behaviour [here](https://github.com/openfaas/faas-cli/blob/f60910aea074e0ab5e5ada880f426d85fa9d4cb7/versioncontrol/git.go#L18).

**Comment by iyovcheva** [comment]
@johnmccabe it's set to `true`

**Comment by johnmccabe** [comment]
@ImVexed can you do the same please (just to be sure)

@iyovcheva ok so that looks like its the root cause here, you could try setting this to false locally [here](https://github.com/openfaas/faas/issues/653#issuecomment-384339538), and trying with a fresh function. That should prevent this reoccurring in future I suspect. (my Win10 box is at home so can't verify this myself until later).

**Comment by ImVexed** [comment]
@johnmccabe can confirm over here. Setting it to `false` indeed fixed the issue.

**Comment by alexellis** [comment]
If this is the work-around / fix - then what would you all suggest we do to move forward? Document this or keep working on an alternative solution?

**Comment by alexellis** [comment]
Please feel free to suggest an edit to the site at docs.openfaas.com for the CLI. https://github.com/openfaas/docs @ImVexed
